src/data/tree.lean

```diff
@@ -81,4 +84,64 @@ def map {β} (f : α → β) : tree α → tree β
 | nil := nil
 | (node a l r) := node (f a) (map l) (map r)
 
+/-- The number of internal nodes (i.e. not including leaves) of a binary tree -/
+@[simp] def num_nodes : tree α → ℕ
+| nil := 0
+| (node _ a b) := a.num_nodes + b.num_nodes + 1
+
+/-- The number of leaves of a binary tree -/
+@[simp] def num_leaves : tree α → ℕ
+| nil := 1
+| (node _ a b) := a.num_leaves + b.num_leaves
+
+/-- The height - length of the longest path from the root - of a binary tree -/
+@[simp] def height : tree α → ℕ
+| nil := 0
+| (node _ a b) := max a.height b.height + 1
+
+lemma num_leaves_eq_num_nodes_succ (x : tree α) : x.num_leaves = x.num_nodes + 1 :=
+by { induction x; simp [*, nat.add_comm, nat.add_assoc, nat.add_left_comm], }
+
+lemma num_leaves_pos (x : tree α) : 0 < x.num_leaves :=
+by { rw num_leaves_eq_num_nodes_succ, exact x.num_nodes.zero_lt_succ, }
+
+lemma height_le_num_nodes : ∀ (x : tree α), x.height ≤ x.num_nodes
+| nil := le_refl _
+| (node _ a b) := nat.succ_le_succ
+    (max_le
+      (trans a.height_le_num_nodes $ a.num_nodes.le_add_right _)
+      (trans b.height_le_num_nodes $ b.num_nodes.le_add_left _))
+
+/-- The left child of the tree, or `nil` if the tree is `nil` -/
+@[simp] def left : tree α → tree α
+| nil := nil
+| (node _ l r) := l
+
+/-- The right child of the tree, or `nil` if the tree is `nil` -/
+@[simp] def right : tree α → tree α
+| nil := nil
+| (node _ l r) := r
+
+/-- Abbreviation for making a node with `unit` data -/
+@[pattern] abbreviation unode : tree unit → tree unit → tree unit :=
+tree.node ()
```

Comment:
Eric suggests that you use localized notation instead of an abbreviation. And I now recall that @b-mehta introduced exactly this in his local formalisation of the Seven Trees in One paper.

Reply:
What notation do you suggest? Should it just be the name `unode`? I'm unsure of the name here anyway.

Reply:
I ended up using the notation "△" defined in a new locale "tree." The "▲" was not already defined in the VS Code's extension's keybinding map/translation map, and I couldn't figure out if there was a mathlib file that adds extra preferences for the extensions or defines additional custom translations. (I can do it locally, but I don't think it is a good idea to have everyone have to update their extension preferences).

Reply:
The list is here: https://github.com/leanprover/vscode-lean/blob/master/src/abbreviation/abbreviations.json

Reply:
Is there anything wrong with this notation? Easier to use existing notation than to make a PR on a different repository.

Reply:
No, sure. I just don't want you to limit yourself because you don't know where things are. Adding new notation to the extension is really not that hard, because you don't need the shortcut to start using it.